### ISSUE TITLE
Context.WithTimeout not stopping command or how do I stop a command 

### ISSUE THREAD

**Issue by lpegoraro** [spf13/cobra#1868]
Hey, I was doing some benchmark test on a app, and I wanted to do e2e, calling the cobra.Command.ExecuteContext(ctx) and I setup the context with a timeout.
iex: ctx, cancelFunc := context.WithTimeout(context.Background(), 2 * time.Minute) 

When the timeout would hit, I was thinking that the context would be cancelling, and stopping the execution.
1. Is that the expected behavior? 
2. If not how can I stop a execution of the Run command after it started?

**Comment by johnSchnake** [comment]
I'm not sure how familiar you are with the context timeouts in general but it doesnt interrupt execution automatically. You have to be checking a channel to see if you've had the context be cancelled/timeout for any reason. There are a number of resources you can find to show how this is done like [here](https://medium.com/geekculture/timeout-context-in-go-e88af0abd08d#:~:text=In%20concurrent%20programming%20with%20Golang%2C%20the%20context%20package%20is%20a,request%20or%20a%20database%20request).

**Comment by bilalcaliskan** [comment]
hello @johnSchnake , i am listening Done channel but still does not get timeout somehow. here is the example:

```
RunE: func(cmd *cobra.Command, args []string) error {
	ctx, cancel := context.WithTimeout(cmd.Context(), 10*time.Second)
	defer cancel()
	cmd.SetContext(ctx)

	for i := 0; i < 20; i++ {
		time.Sleep(2 * time.Second)
		fmt.Println("hello friend, i am dancing right here")
	}

	select {
	case <-ctx.Done():
		fmt.Printf("Context cancelled: %v\n", ctx.Err())
	}
```

**Comment by lpegoraro** [comment]
@johnSchnake, yeah I understand and completely following up that I would be listening to the done of the context by a similar select that @bilalcaliskan exampled.

**Comment by bilalcaliskan** [comment]
@lpegoraro somehow this select is not working on me with `SetContext` method but please let us now when you try that.

**Comment by pantelis-karamolegkos** [comment]
The fact that you have to call `defer cancel()` once the context is set, doesn't make the usage of global command timeouts set in the `rootCmd` unusable?

I am asking because I am setting it in my `rootCmd` as follows


```go
ctxInit := context.Background()
timeout := viper.GetInt("timeout")
ctx, cancel := context.WithTimeout(ctxInit, time.Duration(timeout)*time.Second)
defer cancel()
```

then in a subcommand

```go
ctx := rootCmd.Context()
```

but `ctx` is `context.emptyCtx {}`

(unless I am passing it the wrong way)

**Comment by marckhouzam** [comment]
@pantelis-karamolegkos my guess is that the problem is the `differ cancel ()`. This gets called as soon as the function in which it is called completes. The way you describe your situation. I think this gets called before the sub command.
